### PR TITLE
merging: add metrics for shard count and janitor jobs

### PR DESCRIPTION
**Pull request by stefanhengl**
This change gives us real time metrics for compound shard count, vacuuming and merging.

Co-authored-by: geoffrey@sourcegraph.com

### CODE REVIEW

File: cmd/zoekt-sourcegraph-indexserver/main.go
Comment (tsenart) [inline cmd/zoekt-sourcegraph-indexserver/main.go]: nit: What does `fns` stand for here?

Comment (tsenart) [reply cmd/zoekt-sourcegraph-indexserver/main.go]: I think we could use https://pkg.go.dev/io/fs#Glob here.

File: cmd/zoekt-sourcegraph-indexserver/merge.go
Comment (tsenart) [inline cmd/zoekt-sourcegraph-indexserver/merge.go]: How about we also capture how long merging takes, together with if it errored or not?

File: cmd/zoekt-sourcegraph-indexserver/main.go
Comment (stefanhengl) [reply cmd/zoekt-sourcegraph-indexserver/main.go]: `fns` stands for filenames, we use it in cleanup.go, too.

Comment (stefanhengl) [reply cmd/zoekt-sourcegraph-indexserver/main.go]: Glob sounds good. I chose filepath.Glob over fs.Glob because we already use it in watcher.go and other places. Any reason to prefer fs.Glob here?

File: cmd/zoekt-sourcegraph-indexserver/merge.go
Comment (ggilmore) [inline cmd/zoekt-sourcegraph-indexserver/merge.go]: Not blocking on this, but ideally we'd want to have a _separate_ metric for the duration of failed merging operations

Comment (stefanhengl) [reply cmd/zoekt-sourcegraph-indexserver/merge.go]: We don't really expect that many failed merges. It should be a rare event. According to our log, there were 0 errors during merging over last 30 days. I think the error counter already gives us a good enough signal for now.

Comment (tsenart) [reply cmd/zoekt-sourcegraph-indexserver/merge.go]: Even if we don't have failures so far (great!), I think we should standardize on always having a normalized `error` label in latency histograms.

File: cmd/zoekt-sourcegraph-indexserver/main.go
Comment (tsenart) [inline cmd/zoekt-sourcegraph-indexserver/main.go]: I was reading the code and I don't think there's a big difference between io/fs.Glob and path/filepath.Glob. The former was introduced more recently, so it's the `io.FS` aware version of Glob, with the extra argument.

I think you can do the filtering of the compound- prefix here already.

```suggestion
	fns, err := filepath.Glob(filepath.Join(indexDir, "compound-*.zoekt"))
```

Comment (tsenart) [reply cmd/zoekt-sourcegraph-indexserver/main.go]: And then there's no need to loop over the fns: `metricNumberCompoundShards.Set(float64(len(fns)))`

**Review by tsenart — APPROVED** [review]
Left some more feedback, approving to unblock.

File: cmd/zoekt-sourcegraph-indexserver/merge.go
Comment (stefanhengl) [reply cmd/zoekt-sourcegraph-indexserver/merge.go]: done

**Pull request merged**
